File: typescript/grains/grains.ts
export const square = (num: number): bigint => {
  if (num < 1 || num > 64) throw new Error('Invalid square');

  return num === 1
    ? 1n
    : 2n * square(num - 1);
}

export const total = (): bigint => {
  let sum = 0n;
  for (let i = 1; i <= 64; i++) {
    sum += square(i);
  }

  return sum;
}

File: c/grains/grains.c
#include "grains.h"

uint64_t square(uint8_t index)
{
    if (index < 1 || index > 64) return 0;
    if (index == 1) return 1;
    return 2 * square(index - 1);
}

uint64_t total()
{
    uint64_t sum = 0;
    for (int index = 1; index <= 64; index += 1)
    {
        sum += square(index);
    }
    return sum;
}
